### PR TITLE
Replace usage of "addslashes" with prepared statements

### LINKED ISSUE #13318: When postgres is used, the setup fails if username contains a dash
### Steps to reproduce
1. Setup Postgres (10.5, 10.6, 11.1 - doesn't matter)
2. Setup Nextcloud 15.0.0
3. Point your browser to your domain (cloud.example.com), which shows the setup dialogue.
4. As username for Nextcloud enter something with a dash (e.g. "Kopf-IT"). Chose PostgreSQL as database type, use postgres as user (the default superuser!) and its password. Fill out everything else as usual

### Expected behaviour
Nextcloud should then use the postgres superuser to create a Nextcloud only user, insert the data into the config.php, and continue to setup the Nextcloud instance.

### Actual behaviour
Nextcloud tries to create the oc_kopf-it user, but fails. The error I get is:
`Error while trying to create admin user: Failed to connect to the database: An exception occured in driver: SQLSTATE[08006] [7[ FATAL: password authentication failed for uer "oc_kopf-it"`

Using KopfIT or any other name not containing a dash circumvents the problem. I know this sounds like a Postgres problem (or a problem Postgres should solve, not Nextcloud), but after searching for similar error message on The Internet™ I came to the conclusion it might be an error in the Nextcloud postgres driver. It seems the username postgres tries to create is not escaped in the right way.

### Server configuration

**Operating system**: Ubuntu 18.04.1 LTS

**Web server:** Caddy

**Database:** Postgres

**PHP version:** 7.2.13

**Nextcloud version:** 15.0.0

**Updated from an older Nextcloud/ownCloud or fresh install:** Fresh install, using the docker image

**Where did you install Nextcloud from:** [Docker Hub](https://hub.docker.com/_/nextcloud)

### PR DESCRIPTION
**Pull request by tianon**
As referenced in https://github.com/nextcloud/server/pull/15187#discussion_r312677641, https://github.com/nextcloud/server/issues/11311, https://github.com/nextcloud/server/issues/1260.

I figured opening a PR was a better (concrete) way to open a discussion around this. :smile: :heart:

I'm not actually a user of nextcloud, so I'm not 100% sure what to do to test this effectively, but I'm willing to give it a shot (although it appears there might be CI that does so already?).

### CODE REVIEW

**Comment by tianon** [comment]
I guess `Error while trying to create admin user: Failed to connect to the database: An exception occurred in driver: SQLSTATE[08006] [7] FATAL: password authentication failed for user "oc_admin"
13` is the answer to why this is using `addslashes` instead of prepared statements, although `addslashes` isn't exactly right either. :disappointed:

**Comment by kesselb** [comment]
> I guess `Error while trying to create admin user: Failed to connect to the database: An exception occurred in driver: SQLSTATE[08006] [7] FATAL: password authentication failed for user "oc_admin" 13`

:pass should be in single quotes (check the examples here: https://www.postgresql.org/docs/9.6/sql-createrole.html).

File: lib/private/Setup/PostgreSQL.php [hunk 1]
```diff
@@ -152,11 +152,11 @@ private function createDBUser(IDBConnection $connection) {
 			}
 
 			// create the user
-			$query = $connection->prepare("CREATE USER " . addslashes($this->dbUser) . " CREATEDB PASSWORD '" . addslashes($this->dbPassword) . "'");
-			$query->execute();
+			$query = $connection->prepare('CREATE USER :user CREATEDB PASSWORD :pass');
```
Comment (kesselb) [inline lib/private/Setup/PostgreSQL.php]:
```suggestion
			$query = $connection->prepare('CREATE USER :user CREATEDB PASSWORD \':pass\'');
```

or with double quotes for the sql statment and single quotes for pass.

Comment (tianon) [reply lib/private/Setup/PostgreSQL.php]: It's a prepared statement, so the quotes are added automatically.

I believe the issue is that the username is getting quoted like a normal string but actually requires a different type of quoting.

Comment (kesselb) [reply lib/private/Setup/PostgreSQL.php]: > It's a prepared statement, so the quotes are added automatically.

Right ;)

Comment (tianon) [reply lib/private/Setup/PostgreSQL.php]: I think https://stackoverflow.com/a/18901826/433558 is the answer; I'll switch `:user` out for `pg_escape_identifier` as soon as I get a chance.

**Comment by kesselb** [comment]
Ref https://github.com/nextcloud/server/issues/13318

**Comment by kesselb** [comment]
https://stackoverflow.com/questions/10216784/what-is-a-valid-postgresql-database-name
https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

cc @nickvergessen @rullzer any idea how to do this properly?

**Comment by rullzer** [comment]
@icewind1991 as you use postgres reguraly

**Comment by kesselb** [comment]
Hello @icewind1991 ;)

**Comment by skjnldsv** [comment]
PIng!

**Comment by J0WI** [comment]
Could anyone else help out here if @icewind1991 is not available? @nextcloud/server-triage

**Comment by blizzz** [comment]
what's the state here? Should it be moved to 23?

**Comment by tianon** [comment]
It really would benefit from someone more well-versed in both nextcloud and
postgresql taking over and figuring out the full proper fix.

**Pull request closed without merging**